### PR TITLE
Introduce media type to recordings to allow filtering between recordings and VOD

### PR DESCRIPTION
**Pull request by phunkyfish**
## Description

Introduce a media type enum to PVR recordings to be able to distinguish PVR Recordings. The PR introduce VOD in addition to standard PVR recordings.

## Motivation and context

If a user add-on has VOD in their PVR add-on to be able to filter these in the UI. This PR is just the API change required to implement filtering.

Filtering could be done in the current recordings UI, or even as a separate tile, I'm open to suggestions here.

## How has this been tested?

Using iptvsimple on mac.

## What is the effect on users?

None yet.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h [hunk 1]
```diff
@@ -424,6 +426,25 @@ class PVRRecording : public CStructHdl<PVRRecording, PVR_RECORDING>
   /// @brief To get with @ref SetChannelType changed values
   PVR_RECORDING_CHANNEL_TYPE GetChannelType() const { return m_cStructure->channelType; }
 
+  /// @brief **optional**\n
+  /// Channel type.
+  ///
+  /// Set to @ref PVR_RECORDING_MEDIA_TYPE_UNKNOWN if the type cannot be
```
Comment (ksooo) [inline xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h:432]: The new enum has no such value. And the whole implementation is missing support for this case.

File: xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h [hunk 2]
```diff
@@ -424,6 +426,25 @@ class PVRRecording : public CStructHdl<PVRRecording, PVR_RECORDING>
   /// @brief To get with @ref SetChannelType changed values
   PVR_RECORDING_CHANNEL_TYPE GetChannelType() const { return m_cStructure->channelType; }
 
+  /// @brief **optional**\n
+  /// Channel type.
```
Comment (ksooo) [inline xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h:430]: Recording media type?

File: xbmc/pvr/recordings/PVRRecording.h [hunk 3]
```diff
@@ -32,6 +32,12 @@ class CPVRClient;
 class CPVRProvider;
 class CPVRTimerInfoTag;
 
+enum class RecordingMediaType : bool
+{
+  PVR_RECORDING_MEDIA_TYPE_RECORDING,
+  PVR_RECORDING_MEDIA_TYPE_VOD
+};
```
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h:39]: PVR_RECORDING_MEDIA_TYPE_UNKNOWN and implementation missing.

File: xbmc/pvr/recordings/PVRRecording.h [hunk 4]
```diff
@@ -239,6 +245,12 @@ class CPVRRecording final : public CVideoInfoTag
    */
   bool IsRadio() const { return m_bRadio; }
 
+  /*!
+   * @brief Retrieve the type of media this recording is
+   * @return PVR_RECORDING_MEDIA_TYPE_RECORDING if this is a pvr recording, PVR_RECORDING_MEDIA_TYPE_VOD if this is a tv recording
```
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h:250]: What about PVR_RECORDING_MEDIA_TYPE_UNKNOWN ?

File: xbmc/pvr/recordings/PVRRecording.h [hunk 5]
```diff
@@ -535,6 +547,7 @@ class CPVRRecording final : public CVideoInfoTag
   std::string m_strProviderName; /*!< name of the provider this recording is from */
   int m_iClientProviderUniqueId =
       PVR_PROVIDER_INVALID_UID; /*!< provider uid associated with this recording on the client */
+  RecordingMediaType m_mediaType; /*!< The type of media this recording is */
```
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h:550]: Should be initialized with PVR_RECORDING_MEDIA_TYPE_UNKNOWN.

**Review by ksooo — COMMENTED** [review]
Not sure if the "unknown" value makes sense, as implementation should always be able to determine the type, right? Otherwise, implementation for the "unknown" case is missing.

**Comment by phunkyfish** [comment]
> Not sure if the "unknown" value makes sense, as implementation should always be able to determine the type, right? Otherwise, implementation for the "unknown" case is missing.

I agree. We should not have an unknown type. Let me clean up the commented reference for that mum value.

File: xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h (hunk 1 above)
Comment (phunkyfish) [reply xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h:432]: Fix. API states that media type should always be known to the client and the default value will be PVR_RECORDING_MEDIA_TYPE_RECORDING

File: xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h (hunk 2 above)
Comment (phunkyfish) [reply xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h:430]: Fixed

File: xbmc/pvr/recordings/PVRRecording.h (hunk 3 above)
Comment (phunkyfish) [reply xbmc/pvr/recordings/PVRRecording.h:39]: Not relevenat, media type will always be known.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 4 above)
Comment (phunkyfish) [reply xbmc/pvr/recordings/PVRRecording.h:250]: Not relevenat, media type will always be known.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 5 above)
Comment (phunkyfish) [reply xbmc/pvr/recordings/PVRRecording.h:550]: Not relevenat, media type will always be known.

File: xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h (hunk 1 above)
Comment (ksooo) [reply xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h:432]: Perfect.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 5 above)
Comment (ksooo) [reply xbmc/pvr/recordings/PVRRecording.h:550]: anyway, every member should be initialized,  with TYPE_RECORDING in this case

Comment (phunkyfish) [reply xbmc/pvr/recordings/PVRRecording.h:550]: True, I've added that default value.

File: xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/pvr/pvr_recordings.h [hunk 6]
```diff
@@ -99,6 +99,26 @@ extern "C"
...
+  /// @ingroup cpp_kodi_addon_pvr_Defs_Recording
+  /// @brief **PVR recording media types**\n
+  /// Used on @ref kodi::addon::PVRRecording::SetMediaType() value to set related
+  /// type.
+  ///
+  ///@{
+  typedef enum PVR_RECORDING_MEDIA_TYPE
+  {
+    /// @brief __0__ : Unknown type.
+    PVR_RECORDING_MEDIA_TYPE_RECORDING = 0,
```
Comment (ksooo) [inline xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/pvr/pvr_recordings.h:113]: Description does not match semantics.

File: xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/pvr/pvr_recordings.h [hunk 7]
```diff
@@ -99,6 +99,26 @@ extern "C"
...
+
+    /// @brief __1__ : TV channel.
+    PVR_RECORDING_MEDIA_TYPE_VOD = 1,
```
Comment (ksooo) [inline xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/pvr/pvr_recordings.h:116]: Description does not match semantics?

File: xbmc/pvr/recordings/PVRRecording.cpp [hunk 8]
```diff
@@ -145,6 +145,10 @@ CPVRRecording::CPVRRecording(const PVR_RECORDING& recording, unsigned int iClien
       }
     }
   }
+  if (recording.mediaType == PVR_RECORDING_MEDIA_TYPE_RECORDING)
+    m_mediaType = RecordingMediaType::PVR_RECORDING_MEDIA_TYPE_RECORDING;
+  else
+    m_mediaType = RecordingMediaType::PVR_RECORDING_MEDIA_TYPE_VOD;
```
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.cpp:151]: To be safe against missing future extensions of the add-on API enum here, I would check all known values and add a final else with an error/warning log that we're facing an unknown/unhandled value here.

```c++
else if (recording.mediaType == PVR_RECORDING_MEDIA_TYPE_VOD)
    m_mediaType = RecordingMediaType::PVR_RECORDING_MEDIA_TYPE_VOD;
else
{
    Clog::Log(..., "Unhandled enum value, defaulting to PVR_RECORDING_MEDIA_TYPE_RECORDING");
    m_mediaType = RecordingMediaType::PVR_RECORDING_MEDIA_TYPE_RECORDING;
}
```

File: xbmc/pvr/recordings/PVRRecording.h [hunk 9]
```diff
@@ -32,6 +32,12 @@ class CPVRClient;
 class CPVRProvider;
 class CPVRTimerInfoTag;
 
+enum class RecordingMediaType : bool
```
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h:35]: bool ?

File: xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/pvr/pvr_recordings.h (hunk 6 above)
Comment (phunkyfish) [reply xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/pvr/pvr_recordings.h:113]: Done

File: xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/pvr/pvr_recordings.h (hunk 7 above)
Comment (phunkyfish) [reply xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/pvr/pvr_recordings.h:116]: Done

File: xbmc/pvr/recordings/PVRRecording.cpp (hunk 8 above)
Comment (phunkyfish) [reply xbmc/pvr/recordings/PVRRecording.cpp:151]: Great idea. Added, I also added it to the fill data call.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 9 above)
Comment (phunkyfish) [reply xbmc/pvr/recordings/PVRRecording.h:35]: Done

**Comment by flubshi** [comment]
I think it is a pretty good idea to improve the VoD hadling of Kodi. Some thoughts when thinking about the implementation of my pvr addon:


**VoD (Recording) vs Playable EPG Tag vs Recordings**

In case of this pvr addon, there are live TV channels and VoD channels. VoD channels at the first moment look like a normal channels, except that every EPG Tag for this channel in the past and the future is playable. From my pov, these EPG Tags should be handled the same way as VoD entries expose via the recording API.

This TV provider also provides a dedicated VoD interface (besides VoD channels). This VoD interface uses a separate API than the recordings API. Not sure how other providers handle this. But the question arises, whether extending the recordings API is the best solution to provide a VoD interface.
If we use the recordings API we cannot, for instance, update the recordings separate from the VoD database. In some cases a different update interval might be useful, especially if there is a larger VoD offer.

In addition, there are differences between VoD and recordings. For instance, recordings can be deleted from backend while VoD ususally not. I think, this also needs to be addressed when integrating VoD in recordings.


**VoD UI**

When designing a VoD UI, we should keep the user relevance of the entries in mind. For a user, a manual programmed recording might be a more relevant than the mass of VoD entries. Especially when there a lot of VoD entries, the more relevant recordings might not be fast accessible by the user. Maybe it makes sense to make a separate UI for VoD?

For TV shows and Movies, Kodi already has a separate menu. Maybe it makes sense to integration Movies and TV shows from the VoD offer in this menu? But it feels like a duplicate of the [Media importing and library integration](https://forum.kodi.tv/showthread.php?tid=224794), created by @Montellese 

For VoD entries (except movies and TV shows), we would still need a solution. I am undecided wheter it makes sense to attach them to the channels or group the by categories (news, magazines, ...).

File: xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h [hunk 10]
```diff
@@ -424,6 +426,28 @@ class PVRRecording : public CStructHdl<PVRRecording, PVR_RECORDING>
...
+  /// all availble values. Media Type should always be known to the client.
+  ///
+  /// --------------------------------------------------------------------------
+  ///
+  /// @copydetails cpp_kodi_addon_pvr_Defs_Recording_PVR_RECORDING_MEDIA_TYPE
+  ///
+  /// Example:
+  /// ~~~~~~~~~~~~~{.cpp}
+  /// kodi::addon::PVRRecording tag;
+  /// tag.SetChannelType(PVR_RECORDING_MEDIA_TYPE_RECORDING);
```
Comment (ksooo) [inline xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h:443]: tag.SetMediaType(...

File: xbmc/pvr/recordings/PVRRecording.h (hunk 4 above)
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h:250]: Sorry, what is a "tv recording? Do you mean "VOD recording"? Not sure.

File: xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h (hunk 10 above)
Comment (phunkyfish) [reply xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/pvr/Recordings.h:443]: Fixed

File: xbmc/pvr/recordings/PVRRecording.h (hunk 4 above)
Comment (phunkyfish) [reply xbmc/pvr/recordings/PVRRecording.h:250]: Fixed

**Comment by phunkyfish** [comment]
> Media importing and library integration

The PVR add-on should decide the best way to show it's media to the user for now until Kodi itself is more opinionated. Whether that be via catch/archive through channels/EPG or via the recordings interface. Unifying EPG and Recordings is way out of scope right now. This PR is a start on making things a little more compatible, and to start defining the difference between recordings and VOD.

When media import lands there will be a lot more options, but who knows when that will happen. Also, we should keep discussions like this to the the forum, and not on PRs.

**Comment by emveepee** [comment]
Has this been stress tested?  Many of the "subscriber" IPTV users who we know will end up using this feature probably have 20000+ VOD recordings and some users m3u files I have seen had 100,000+.  It's tough enough to get users to slim down the number of channels to avoid timeouts and I don't want to create a support nightmare for myself.

I think many of the capabilities for VOD need to be configurable separate from recordings too.  From the NextPVR backend size, rename, delete and undelete won't be available. Some capabilities with vary by VOD type too I know for one of my NextPVR input addons that has VOD I won't be able to support LastPlayed and  skipping forward won't be possible until that part of the video is downloaded.  Not sure if any providers have EDL support.

File: xbmc/pvr/recordings/PVRRecordingsPath.cpp [hunk 11]
```diff
@@ -32,10 +33,12 @@ CPVRRecordingsPath::CPVRRecordingsPath(const std::string& strPath)
...
-  m_bValid = ((segments.size() >= 4) && // at least pvr://recordings/[tv|radio]/[active|deleted]
-              StringUtils::StartsWith(strVarPath, "pvr://") && (segments.at(1) == "recordings") &&
-              ((segments.at(2) == "tv") || (segments.at(2) == "radio")) &&
-              ((segments.at(3) == "active") || (segments.at(3) == "deleted")));
+  m_bValid =
+      ((segments.size() >= 4) && // at least pvr://recordings/[tv|radio]/[active|deleted]/[rec|vod]
+       StringUtils::StartsWith(strVarPath, "pvr://") && (segments.at(1) == "recordings") &&
+       ((segments.at(2) == "tv") || (segments.at(2) == "radio")) &&
+       ((segments.at(3) == "active") || (segments.at(3) == "deleted")) &&
+       ((segments.at(4) == "rec") || (segments.at(3) == "vod")));
```
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecordingsPath.cpp:41]: This line reads wrong. Both indexes should be 4 here, right?

**Comment by ksooo** [comment]
> Has this been stress tested? Many of the "subscriber" IPTV users who we know will end up using this feature probably have 20000+ VOD recordings and some users m3u files I have seen had 100,000+. 

This is a very valid point. Keeping everything in memory, could easily lead to resource problems. Look at what we have learned from EPG , where we once switched from all in memory to read on demand (from db) approach.

**Comment by phunkyfish** [comment]
This is much further down the line than where we are right now. VOD has been supported with pvr.iptvsimple for a full release now though and many users have hundreds of thousands of VODs. No users complaints from users so far 😉

And I  take the point we need to trim down the functionality for VOD, but that will be a later commit.

**Comment by emveepee** [comment]
> This is much further down the line than where we are right now. VOD has been supported with pvr.iptvsimple for a full release now though and many users have hundreds of thousands of VODs. No users complaints from users so far 😉

OK using the NextPVR logic VOD is currently only a channel and even as a db the default from the backend is not to send VOD as a channel because of many user complaints about load time and navigation in the guide.    I'll have to fake some large m3u lists to see how they perform, I am not an IPTV subscriber. 

Perhaps another solution would be to for core to handle the channel to VOD conversion that currently is done as media conversion in pvr.iptvsimple code, if recordings are more efficient.   I'd have to think about loading other metadata like descriptions.

**Comment by emveepee** [comment]
> 100K channels is definitely not a good idea. PVR core is not designed to handle so many channels in an efficient way, sorry Changing this would be huge effort and is nothing on my todo list. Channel != VOD! At best Channel = sequence of VOD, though also not a clean approach. Imo, mapping a single VOD to a channel is just bad hack.

The issue is the IPTV m3u file is overloaded so all IPTV handlers have to deal with it.  If you look at pvr.iptvsimple https://github.com/kodi-pvr/pvr.iptvsimple/blob/Omega/src/iptvsimple/PlaylistLoader.cpp#L224 the channel and VOD are in the same playlist.  I was suggesting that a field on a the channel could do that logical VOD split in core rather than addons repeating all the media logic from pvr.iptvsimple.

**Comment by ksooo** [comment]
> The issue is the IPTV m3u file is overloaded so all IPTV handlers have to deal with it.

Users/providers can always come up with ridiculous number of channels. All I want to say is that they should not complain about such nonsense use cases. Nobody can handle to watch all of these channels. I will not invest a single hour of my free time to make those "channel collectors" happy.

**Comment by ksooo** [comment]
>  I was suggesting that a field on a the channel could do that logical VOD split in core rather than addons repeating all the media logic from pvr.iptvsimple.

No, I still think to use PVR_CHANNEL to transport a single VOD is the wrong approach. It is semantically not correct. A channel simply is not a VOD. A channel is an endless stream, where a VOD is a video with a start and an end.

If at all, then it should be done like in this PVR, because recordings and VOD are semantically at least related, both are video files.

**Comment by phunkyfish** [comment]
The overloaded nature of M3Us for VODs and its use in iptvsimple should be removed from this conversation. We can’t change that that, it’s an informal standard now. It’s just a delivery mechanism for metadata. It did provide however some proof that handing VODs separately yields huge benefits and preserves the use of channels for its intended purpose. 

We have a whole dev cycle now to do something specific for VODs that makes more sense. And we are in no rush to figure that out. We can take our time. And if we choose, just implement the small API change and move the feature impl to a separate future PR.

File: xbmc/pvr/recordings/PVRRecordingsPath.cpp (hunk 11 above)
Comment (phunkyfish) [reply xbmc/pvr/recordings/PVRRecordingsPath.cpp]: Fixed

**Comment by ksooo** [comment]
> It did provide however some proof that handing VODs separately yields huge benefits and preserves the use of channels for its intended purpose.

I could not agree more.

**Comment by emveepee** [comment]
> The overloaded nature of M3Us for VODs and its use in iptvsimple should be removed from this conversation. We can’t change that that, it’s an informal standard now. It’s just a delivery mechanism for metadata. It did provide however some proof that handing VODs separately yields huge benefits and preserves the use of channels for its intended purpose.

I agree with handling them separately, NextPVR does offer a VOD list api so that isn't the issue. I was noting the the backend will most like populate VOD during a new channel scan so they are tied to the channel triggers not recording triggers so they aren't recordings either.  

I was not saying GetChannels() should populate VOD.   Because the GetRecordings() calls can be frequent, especially in prime time when users don't like too many heavy calls to the backend,  pvr.nextpvr will need to cache them during GetChannels which I feel is the only logical place.  This means that these potentially large lists are in memory twice.    

If pvr.nextpvr could trigger GetRecordings() with VOD and all other GetRecording() calls from the backend are without VOD that would make more sense to me.  Sort of like GetChannels() for radio and TV can be independent.   I guess this is a good point to add there probably needs to be an AOD media option for podcasts and radio shows even if pvr.nextpvr can't handle that.

**Comment by ksooo** [comment]
Here are my thoughts: I think VODs are neither recordings nor programme entries of a channel. Thus, VODs should be designed and implemented as a separate new PVR subcomponent, just like recordings, like channels. 

At the API level, this would require a new struct PVR_VOD and "surrounding" functions, like `GetVODs()`, the respective addon capability `SUPPORTS_VOD`, `TriggerVODsUpdate()`etc, pretty much the same we have for the other subcomponents.

In the Kodi GUIthis would lead to a new window for VODs, new Home screen tile for "VOD" etc.



I know this is much more work than "hacking" VODs into recordings or channels, but imo this is the way to go wrt clean design.

**Comment by phunkyfish** [comment]
> Here are my thoughts: I think VODs are neither recordings nor programme entries of a channel, nor channels. Thus, VODs should be designed and implemented as a separate new PVR subcomponent, just like recordings, like channels.
> 
> At the API level, this would require a new struct `PVR_VOD` and "surrounding" functions, like `GetVODs()`, the respective addon capability `SUPPORTS_VOD`, `TriggerVODsUpdate()` etc., pretty much the same we have for the other subcomponents.
> 
> In PVR core this will result in a new PVR Manager subcomponent `VODs()`.
> 
> In the Kodi GUI, this would lead to a new window for VODs, new Home screen tile for "VOD" etc.
> 
> I know this is much more work than "hacking" VODs into recordings or channels, but imo this is the way to go wrt clean design.
> 
> As @phunkyfish said, we are not in a rush. No need for quick and dirty tricks. We could do it "right" from the beginning for P.

As it happens I implemented this previously. But someone rejected that PR. Let me dig if up 😉

**Comment by phunkyfish** [comment]
Yup, here it is: https://github.com/phunkyfish/xbmc/commit/7d234065dfc35aa60134e01e4402ff1a63433756

I can resurrect it as a PR

**Comment by ksooo** [comment]
> Any preference for what it is called in the UI? VOD makes sense for videos, but not for radio/audio. 

 Oh, fair, but not being a native English speaker, no clue tbh. Maybe we should not talk about "OD" at all and just call it "Media" (or "Other media" to make clear, that this is something additional to other media like channels, recordings etc.)?
 
> Or @ksooo are you suggesting we have TV, Radio and VOD, podcasts etc.?

 I think it should be a subcomponent of PVR and Radio, not another toplevel component. We are already getting questions because for example add-ons like "Radio.de" do not appear under "Radio". Introducing "VOD" as toplevel component, would confuse users even more, I'm afraid ("Why does "Disney+" not appear under "VOD"?)...

**Comment by Long-Man** [comment]
For radio and tv "VOD" it could be called **_Replay Media_** since the expectation is that it was already broadcasted.

**Comment by emveepee** [comment]
I don't like AOD either it was to make a point, perhaps keep it out of scope.  Most audio content is probably better handled by an add-on anyway or defer it to improved m3u handling in core.  Listen On Demand would work though for radio shows, podcast (too bad is wasn't aodcast) and audio books.

As linear grid-based TV fades to on demand, even for broadcaster web sites and apps, maybe Kodi UI needs an overhaul but I'd do that in v23 after much thought.   Kodi pirate builds probably do a better job presenting on demand streaming media than base Kodi.  

@phunkyfish I checked out your commit and I don't think PVR backends should be content and metadata scrapers and aggregators and providers.  There are already UPnP, Jellfyfin, Emby and Plex for those that want that.  Historically NextPVR does provide that but it was inferior to other options.

Your PR is a big stretch from an IPTV list with VOD or PlutoTV VOD stitch URLs that I see feeding into Kodi from the NextPVR backend.     Fundamentally this PR would meet the demands of most current VOD users, I just found the pull mechanism flawed and would prefer push delivery because the addon knows if a list changes better than Kodi, they aren't that dynamic.  

One advantage thing about keeping it in Recordings is the low impact on skinners, it just is another filter type.   I don't even know how many PVR addons support VOD (or are supported at all for that matter)

**Comment by phunkyfish** [comment]
We can go either way. Add it to recordings or have it as its own type. I really don’t mind, but this time around I would like to decide on direction before doing any implementation work.

**Comment by phunkyfish** [comment]
Closing in favour of alternate implementation

**Pull request closed without merging**